### DISCUSSION TITLE
Does filament support casting of values in forms?

### DISCUSSION THREAD

**Discussion by DannyvdSluijs** [filamentphp/filament#3490] [Help] [answered]
First off, great work on filament it has become my goto tool if I just need some quick and easy admin panel.

One thing I'm stuck at is value casting. As some of my Laravel models have a ValueObject cast defined in their resp. `Eloquent` model. This ValueObject can be considered as a wrapper class and I can adjust this if needed. However I don't know how I can.

After looking in the documentation on both the form elements and the page elements I was unable to find any information on such a thing. My filament page is currently break as the value object is rendered in JS as a `Object` and in the console I can see
```
The specified value "[object Object]" cannot be parsed, or is out of range.
```
Saving the form fails as the field is required an the rendering of the page failed before with the above message leaving the values empty.

I did run into the [hooks](https://filamentphp.com/docs/1.x/admin/resources#hooks) but I'm not quite sure if this would be the correct spot to be. Also https://github.com/filamentphp/filament/discussions/2284 seems to be related although the method `afterStateHydrated` cannot be found in my code base (including the vendor folder).

Is there someone that can help? And maybe ultimately if this is supported we could make it clearer in the documentation.

**Comment by danharrin** [comment, marked as answer]
Try these mutate hooks, which you can use to construct and destruct the value objects. The methods need to be put on the respective resource page classes, not the resource itself.

If you need to transform the value objects into an array state before the form is filled:

https://filamentphp.com/docs/2.x/admin/resources/editing-records#customizing-data-before-filling-the-form

If you need to transform them back into a value object when the form is submitted (but I kinda doubt this as maybe your cast handles this anyway?)

https://filamentphp.com/docs/2.x/admin/resources/creating-records#customizing-data-before-saving
https://filamentphp.com/docs/2.x/admin/resources/editing-records#customizing-data-before-saving

**Reply by DannyvdSluijs** [reply]
Thanks for the quick response. You explicitly pointing to version 2 documentation. Is my assumption correct this isn’t available in version 1?

**Reply by danharrin** [reply]
You can use hooks in version 1 to achieve the same thing by mutating `$this->record` data I think, but I would strongly advise upgrading to v2, it's so much more powerful.

<img width="945" alt="Screenshot 2022-08-13 at 15 04 13" src="https://user-images.githubusercontent.com/41773797/184497718-614fc763-1421-4884-b1d0-1364ed130661.png">

I always link to v2 docs by default because we have 160x the number of users using v2 vs v1.
